### PR TITLE
Refactor/init type hints

### PR DESCRIPTION
**Pull request by macintoshpie**
#### Any background context you want to provide?
See linked issues

#### What's this PR do?
- Adds mypy to tox and CI process
- Small refactors to SEED to pass mypy type checking -- some cases I just added some comments to ignore type checking, and in others I removed outdated code
- Adds type hints to seed/models/derived_columns.py for examples

#### How should this be manually tested?
N/A, but definitely look through the edited `derived_columns.py` file

#### What are the relevant tickets?
#3034 

#### Screenshots (if appropriate)
Example of an untyped func with hints (VSCode)
Screen Shot 2021-12-15 at 2 47 05 PM

Example using type hints for derived_columns stuff
Screen Shot 2021-12-15 at 3 29 56 PM

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -105,14 +105,14 @@ jobs:
...
-        tox_env: [flake8, docs]
+        tox_env: [flake8, docs, mypy]
     steps:
       -
         uses: actions/checkout@v2
       -
         uses: actions/setup-python@v2
         with:
-          python-version: '3.7'
+          python-version: '3.9'
```
Comment (macintoshpie) [inline .github/workflows/ci.yml:115]: SEED is on 3.9!

File: mypy.ini [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[mypy]
+ignore_missing_imports = True
+
+[mypy-seed.*.migrations.*]
+ignore_errors = True
```
Comment (macintoshpie) [inline mypy.ini:5]: Ignoring migration files for now b/c they raise type issues that aren't helpful

File: seed/models/meters.py [hunk 3]
```diff
@@ -64,7 +64,7 @@ class Meter(models.Model):
         (ELECTRICITY_UNKNOWN, 'Electric - Unknown'),
     )
 
-    type_lookup = dict((reversed(type) for type in ENERGY_TYPES))
+    type_lookup = dict((reversed(type) for type in ENERGY_TYPES))  # type: ignore
```
Comment (macintoshpie) [inline seed/models/meters.py:67]: Great example of types not working well sometimes. mypy is unable to infer that the Tree has a decorator which evaluates the tree (in this case into a float)

File: seed/models/derived_columns.py [hunk 4]
```diff
@@ -150,7 +152,7 @@ def evaluate(self, parameters=None):
             parameters = {}
 
         self._transformer.set_params(parameters)
-        return self._parser.parse(self._expression)
+        return self._parser.parse(self._expression)  # type: ignore[return-value]
```
Comment (macintoshpie) [inline seed/models/derived_columns.py:155]: Great example of types not working well sometimes. mypy is unable to infer that the Tree has a decorator which evaluates the tree (in this case into a float)

File: docs/source/developer_resources.rst [hunk 5]
```diff
@@ -31,6 +31,44 @@ To run flake locally call:
 
     tox -e flake8
 
+Python Type Hints
```
Comment (macintoshpie) [inline docs/source/developer_resources.rst:34]: Looking for feedback on this. Specifically:
- If you don't use VSCode, what do you use? we should check if there's a suitable extension for it
- Is the usage of types clear and sufficient?

Comment (perryr16) [reply docs/source/developer_resources.rst:34]: - I use VSCode
- I think I'm following. Typing is not mandatory but prefered going forward. If I run into issues I can always use `# type ignore`

Comment (Ryoken) [reply docs/source/developer_resources.rst:34]: I use Sublime and it looks like there's a plugin already that ties mypy in: https://github.com/fredcallaway/SublimeLinter-contrib-mypy

**Review by Ryoken — APPROVED** [review]

**Review by haneslinger — APPROVED** [review]

File: docs/source/developer_resources.rst [hunk 6]
```diff
@@ -31,6 +31,44 @@ To run flake locally call:
...
+- If you're getting warnings about runtime errors due to a type name, make sure your IDE is set up to point to an environment with python 3.9
+- If you're wasting time trying to please the type checker, feel free to throw :code:`# type: ignore` on the problematic line (or at the top of the file to ignore all issues for that file)
+
+Type Checking
+*************
+
+CI currently runs static type checking on the codebase using `mypy <http://mypy-lang.org/>`_. For
+your own IDE, we recommend the following extensions:
+
+- VSCode: `Pylance <https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance>`_ (uses Microsoft's Pyright type checking)
```
Comment (nllong) [inline docs/source/developer_resources.rst:63]: 👍

File: seed/data_importer/__init__.py [hunk 7]
```diff
@@ -4,9 +4,3 @@
 :copyright (c) 2014 - 2021, The Regents of the University of California, through Lawrence Berkeley National Laboratory (subject to receipt of any required approvals from the U.S. Department of Energy) and contributors. All rights reserved.  # NOQA
 :author
 """
-# monkey-patch to suppress threading error message in python 2.7.3
-# See http://stackoverflow.com/questions/13193278/understand-python-threading-bug
-import sys
-if sys.version_info[:3] == (2, 7, 3):
-    import threading
-    threading._DummyThread._Thread__stop = lambda x: 42
```
Comment (nllong) [inline seed/data_importer/__init__.py:12]: yikes, this was in there for a long time!!!

File: seed/data_importer/models.py [hunk 8]
```diff
@@ -11,10 +11,7 @@
 import math
 import tempfile
 
-try:
```
Comment (nllong) [inline seed/data_importer/models.py:14]: cleanup cleanup everybody cleanup

**Review by nllong — APPROVED** [review]
🔥

**Pull request merged**